### PR TITLE
Handle edge case in binary search where a single element vector would…

### PR DESCRIPTION
**Pull request by no-man-is-an-island**
… result in it getting stuck in an infinite loop

We were using the piecewise-linear interpolation function and were (accidentally) passing it a single point to interpolate over. While, admittedly, this was as a result of a bug in our code, it caused the application to hang indefinitely while attempting to do the binary search, and this was very hard to debug.

### CODE REVIEW

**Comment by mikera** [comment]
Seems like a very sensible improvement in terms of behaviour!

Two comments:
- Slight style point, but I think this would be better using the test `(== left right)` as part of a `cond` statement inside the loop. Makes the code a bit shorter, and it avoids the oddness of calling `(count vec)` twice.
- If interpolation with a length 1 vector is always an error, this should probably be picked up and thrown as an exception in the calling code before it even hits the binary search code.

Does that make sense?

**Comment by no-man-is-an-island** [comment]
Definitely agree on your second point - I've committed a check (at the point where the `interpolate` function checks that all the values passed are distinct) that more than one point has been passed.

I agree that checking `(count vec)` twice does look a bit strange, so I've committed a change as you suggest. The reason I did it the way I did was to avoid checking `(== left right)` in every loop (since it can only be true on the first iteration). In practice, I think the performance impact of this will be negligible.

File: modules/incanter-core/src/incanter/interp/utils.clj [hunk 1]
```diff
@@ -2,16 +2,24 @@
...
-          (recur left middle))))))
+    (let [middle (quot (+ left right) 2)]
+      (cond
+       (= right left)
+       0
+
+       (= (- right left) 1)
+       (if (<= (nth vec right) value) right left)
+
+
```
Comment (nblumoe) [inline modules/incanter-core/src/incanter/interp/utils.clj]: Code style: Should be a single empty row only

Comment (no-man-is-an-island) [inline modules/incanter-core/src/incanter/interp/utils.clj]: Thanks!

**Comment by nblumoe** [comment]
Looks good to me, too. Thanks for the PR @no-man-is-an-island 

I was wondering one thing though about the behaviour:

```clojure
incanter.irepl=> (binary-search [1 5 8] 2)
0
incanter.irepl=> (binary-search [3 5 8] 2)
0
```

Is that behavior expected? In the first case the value at index 0 is smaller than second argument, in the other case it is larger, but both calls return the same index.
I don't know the algorithm and intentions for the return values very well, but this seems to be at least in conflict with the docstring ` Finds index of rightmost value in sorted vector that is less or equal to given value.`
If not value is smaller than the given argument, should it maybe return `nil`?

**Comment by no-man-is-an-island** [comment]
I agree @nblumoe - the docstring (and probably the behaviour, too) of `binary-search` is confusing. I also noticed this when implementing the fix for the single-element vector case, but decided to be consistent with the current behaviour.

Looking into it more, the interpolation code that uses this function relies on it returning 0 in the case where all elements in the sequence are greater than the given value. This is to do with what behaviour you want from your interpolation function when you are asked to interpolate to the left of all the points given - the linear interpolation function, for example, will (implicitly) continue to linearly interpolate outside the range. We actually had to work around this behaviour for our use case.

I agree that it would be better if the interpolation function and binary-search function were more explicit about this - you could even allow the caller to specify the behaviour outside the given range of points (some sensible options for points outside the range might be: use left-most and right-most values, continue interpolation using end segments, throw an exception). 

Maybe this is outside the scope of this pull request though?

**Comment by nblumoe** [comment]
@no-man-is-an-island I agree, created a new issue for that https://github.com/incanter/incanter/issues/303

Thanks a lot for the looking into this in your PR! :+1: 

@no-man-is-an-island @mikera I am fine with merging this

**Comment by mikera** [comment]
OK merged!

**Comment by Gerrrr** [comment]
Do we have the same issue in develop branch? If so, shall we merge it also there?

**Pull request merged**